dbms/src/Functions/FunctionsDateTime.h

```diff
@@ -1570,6 +1570,535 @@ class FunctionTimeSlots : public IFunction
+            *target++ = '0' + (day == 7 ? 0 : day);
+        }
+
+        // Year, last two digits (00-99)
+        static void format_y(char *& target, UInt32 source, const DateLUTImpl & timezone)
+        {
+            writeNumber2(target, ToYearImpl::execute(source, timezone) % 100);
+        }
+
+        // Year, 4 digits
+        static void format_Y(char *& target, UInt32 source, const DateLUTImpl & timezone)
+        {
+            writeNumberWidth(target, ToYearImpl::execute(source, timezone), 4);
+        }
+
+        static void format_Percent(char *& target, UInt32 , const DateLUTImpl & )
+        {
+            *target++ = '%';
+        }
+    };
+
+public:
+    static constexpr auto name = "formatDateTime";
+
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionFormatDateTime>(); };
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {1,2}; }
+
+    bool isVariadic() const override { return true; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        if (arguments.size() != 2 && arguments.size() != 3)
+            throw Exception("Number of arguments for function " + getName() + " doesn't match: passed "
+                            + toString(arguments.size()) + ", should be 2 or 3",
+                            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        if (!WhichDataType(arguments[0].type).isDateTime())
+            throw Exception("Illegal type " + arguments[0].type->getName() + " of 1 argument of function " + getName() +
+                            ". Should be a date or a date with time", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        if (!WhichDataType(arguments[1].type).isString())
+            throw Exception("Illegal type " + arguments[1].type->getName() + " of 2 argument of function " + getName() + ". Must be String.",
+                            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        if (arguments.size() == 3)
+        {
+            if (!WhichDataType(arguments[2].type).isString())
+                throw Exception("Illegal type " + arguments[2].type->getName() + " of 3 argument of function " + getName() + ". Must be String.",
+                                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
+    {
+        auto times = checkAndGetColumn<ColumnUInt32>(block.getByPosition(arguments[0]).column.get());
+        auto const_times = checkAndGetColumnConst<ColumnUInt32>(block.getByPosition(arguments[0]).column.get());
+
+        if (const_times)
```

Comment:
It is implemented automatically because of
`useDefaultImplementationForConstants`